### PR TITLE
Support `databuilder --version` on CLI

### LINKED ISSUE #624: Support `databuilder --version` on CLI
Unless anyone has better suggestions, this could be done in the same way as we current do for `opensafely-cli`.

1. Have a file at `databuilder/VERSION` whose default contents is `not-from-a-package` or `development` or something like that. (Bonus points if there's something approximating a convention we can follow here.)

2. Read that file at runtime and wire it up to the `--version` argument.

   https://github.com/opensafely-core/opensafely-cli/blob/a9efc8119a979a16fef1dd9ce1e470adad1be2c2/opensafely/__init__.py#L10
   ```py
   __version__ = Path(__file__).parent.joinpath("VERSION").read_text().strip()
   ```
   https://github.com/opensafely-core/opensafely-cli/blob/a9efc8119a979a16fef1dd9ce1e470adad1be2c2/opensafely/__init__.py#L21-L23
   ```py
       parser.add_argument(
           "--version", action="version", version=f"opensafely {__version__}"
       )
   ```

3. Have the action which builds the package write the correct version to the file before building.
   https://github.com/opensafely-core/opensafely-cli/blob/a9efc8119a979a16fef1dd9ce1e470adad1be2c2/.github/workflows/build_and_publish.yaml#L44-L46
   ```yaml
    - name: Generate correct value for VERSION file
      run: |
        echo ${{ needs.tag-new-version.outputs.tag }} > opensafely/VERSION
   ```

### PR DESCRIPTION
**Pull request by iaindillingham**
The build backend (`setuptools`/`setuptools_scm`) extracts the version from a tag when building the package, and writes the version to the package's metadata following [`setuptools_scm`'s default versioning scheme][1]. (Because we use tags for versioning both software (`vx.y.z`) and metadata (`docs-metadata-<sha>`) we have to override the call to `git describe`.)

The package is installed when creating the image. However, to ensure the tags are available, git is installed and the .git directory is bind-mounted.[^1] There are better ways of building an image from a Dockerfile; implementing them is the subject of #665.

Closes #624

[1]: https://github.com/pypa/setuptools_scm/#default-versioning-scheme

[^1]: https://github.com/pypa/setuptools_scm/#usage-from-docker

### CODE REVIEW

**Comment by ghickman** [comment]
@iaindillingham – I think we have to install the package in the container if we want version numbers here, that's core to how setuptools_scm works.  Given we want to version with git tags (and, I'm assuming, _not_ commit a version file to the repo as part of that workflow) then we have to at the very least build and install.

**Comment by evansd** [comment]
I think we want to install the package in the container in any case. Although it will need the fixes in #671 merged before that will actually work.

There's also ticket #665 open on improving the whole Dockerfile approach based on the one Simon ended up using for cohortextractor.

**Comment by StevenMaude** [comment]
Tagging is also broken at the moment (#651), which may impact this (or at least make it not that useful until fixed).

**Comment by ghickman** [comment]
The Python packaging parts of this look good to me.  Probably best to have someone with more docker knowledge than myself check through the Dockerfile though.

**Review by ghickman — APPROVED** [review]

**Review by bloodearnest — APPROVED** [review]
It would be good to add the version as a docker label too, so it's inspectable from outside easily. e.g. I would like to add the git ref to the metrics I'm currently working on.

I've been trying to add this as GITREF label, e.g. for the python action image:
https://github.com/opensafely-core/python-docker/blob/main/Makefile#L7
https://github.com/opensafely-core/python-docker/blob/main/Dockerfile#L65

Writing this comment has made me realise that the label is GITREF missing in cohortextractor's docker file.

Not blocking, just a suggestion
I've realised this conventions
For the cohortextractor image:

File: .dockerignore [hunk 1]
```diff
@@ -1,5 +1,3 @@
-.git/
```
Comment (evansd) [inline .dockerignore:1]: Don't have time to do a full review right now but, as a quick drive by comment, this doesn't look quite right to me. Doesn't it mean we end up bundling the entire git repo into the Docker image? Which I'm pretty sure we don't want to do.

Comment (bloodearnest) [reply .dockerignore:1]: It's bind mounted at build time on line 33 of the diff

Comment (iaindillingham) [reply .dockerignore:1]: Yes, it's bind-mounted: we need to make the repo available, to extract the version from a tag (see the [_Usage from Docker_][1] section of `setuptools_scm`'s documentation).

[1]: https://github.com/pypa/setuptools_scm/#usage-from-docker

Comment (evansd) [reply .dockerignore:1]: Ah yes. And we never do `COPY .` so it wouldn't get included anyway. Ignore me!

**Comment by iaindillingham** [comment]
You'll find this, buried on line [932][1] of the test failure:

```
  Step 12/13 : RUN --mount=type=bind,target=/app/.git,source=.git python -m pip install --no-cache-dir /app
  ---------------------------- Captured stderr setup -----------------------------
  the --mount option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled
```

The error, on line [396][2], relates to the `databuilder_image` fixture:

https://github.com/opensafely-core/databuilder/blob/e6bd8a47389b114002382108e8f6bbb2fc778a0d/tests/conftest.py#L118-L130

I think I need to add `DOCKER_BUILDKIT=1` as the zeroth element in the list that's passed to `subprocess.run`, to [enable BuildKit builds][3]. Is that correct, @bloodearnest? Or should I enable BuildKit builds everywhere?

[1]: https://github.com/opensafely-core/databuilder/runs/8232454546?check_suite_focus=true#step:5:933
[2]: https://github.com/opensafely-core/databuilder/runs/8232454546?check_suite_focus=true#step:5:397
[3]: https://docs.docker.com/develop/develop-images/build_enhancements/#to-enable-buildkit-builds

**Comment by iaindillingham** [comment]
I've pushed 9e41b5bb and 5a95c133, which build `databuilder` and `databuilder-dev` with BuildKit. I don't know what BuildKit is, but I do know that it allows us to bind-mount the `.git` directory. Could a Docker-maven take a look?

File: .github/workflows/build-and-deploy.yml [hunk 2]
```diff
@@ -98,7 +98,7 @@ jobs:
           echo "MINOR=$MINOR"
           echo "PATCH=$PATCH"
 
-          docker build . --file Dockerfile \
+          DOCKER_BUILDKIT=1 docker build . --file Dockerfile \
```
Comment (bloodearnest) [inline .github/workflows/build-and-deploy.yml:101]: Might be cleaner if you just do

```
export DOCKER_BUILDKIT=1
docker build ...

```

I think that's easier to read, and means if we do other docker builds later.

The way we do this in other projects is have a `just docker-build` command that takes care of any environment variables, and use that in the gh workflow.

Comment (iaindillingham) [reply .github/workflows/build-and-deploy.yml]: Thanks @bloodearnest. I don't want to extend this to editing the workflow too much, but have made the change you suggested in 7fe47369.

**Pull request merged**